jina/drivers/rank/aggregate/__init__.py

```diff
@@ -0,0 +1,211 @@
+                |- chunks: {granularity: k}
+                |    |- matches: {granularity: k}
+                |
+                |- chunks: {granularity: k}
+                    |- matches: {granularity: k}
+        Output:
+        document: {granularity: k-1}
+            |- chunks: {granularity: k}
+            |    |- matches: {granularity: k}
+            |
+            |- chunks: {granularity: k}
+            |    |- matches: {granularity: k}
+            |
+            |-matches: {granularity: k-1} (Ranked according to Ranker Executor)
+
+        .. note::
+            - It traverses on ``chunks`` not on ``matches``. This is because ranker needs context information
+            from ``matches`` for several ``chunks``
+    """
+
+    def __init__(self, traversal_paths: Tuple[str] = ('c',), *args, **kwargs):
+        super().__init__(traversal_paths=traversal_paths, *args, **kwargs)
+
+    def _apply_all(self, docs: 'DocumentSet',
+                   context_doc: 'Document', *args,
+                   **kwargs) -> None:
+        """
+        :param docs: the chunks of the ``context_doc``, they are at depth_level ``k``
+        :param context_doc: the owner of ``docs``, it is at depth_level ``k-1``
+        :return:
+        """
+
+        match_idx = []  # type: List[Tuple[int, int, int, float]]
+        query_meta = {}  # type: Dict[int, Dict]
+        match_meta = {}  # type: Dict[int, Dict]
+        parent_id_chunk_id_map = defaultdict(list)
+        matches_by_id = defaultdict(Document)
+        for chunk in docs:
+            query_meta[int(chunk.id)] = chunk.get_attrs(*self.exec.required_keys)
+            for match in chunk.matches:
+                match_info = self._extract_query_match_info(match=match, query=chunk)
+                match_idx.append(match_info)
+                match_meta[int(match.id)] = match.get_attrs(*self.exec.required_keys)
+                parent_id_chunk_id_map[int(match.parent_id)].append(int(match.id))
+                matches_by_id[int(match.id)] = match
+
+        if match_idx:
+            match_idx = np.array(match_idx,
+                                 dtype=[
+                                     (Chunk2DocRanker.COL_MATCH_PARENT_HASH, np.int64),
+                                     (Chunk2DocRanker.COL_MATCH_HASH, np.int64),
+                                     (Chunk2DocRanker.COL_DOC_CHUNK_HASH, np.int64),
+                                     (Chunk2DocRanker.COL_SCORE, np.float64)
+                                 ]
+                                 )
+
+            docs_scores = self.exec_fn(match_idx, query_meta, match_meta)
+
+            self._insert_query_matches(query=context_doc,
+                                       parent_id_chunk_id_map=parent_id_chunk_id_map,
+                                       chunk_matches_by_id=matches_by_id,
+                                       docs_scores=docs_scores)
+
+
+class CollectMatches2DocRankDriver(BaseAggregateMatchesRanker):
```

Comment:
Not sure about whether it is a better naming. 
```suggestion
class AggregateMatches2DocRankDriver(BaseAggregateMatchesRanker):
```